corgi/tasks/prod_defs.py

```diff
@@ -215,19 +215,46 @@ def parse_product_version(
         _match_and_save_stream_cpes(product_version)
 
 
-def _parse_variants_from_brew_tags(brew_tags) -> list[str]:
+def _parse_variants_from_brew_tags(brew_tags: list[str]) -> dict[str, str]:
     """We don't link core ProductVariant models to streams via brew_tags because it results in too
     many builds being linked to the stream. These variants should include all possible builds that
     could match the stream. We use them for finding which errata match this stream, so it's OK to
     find too many errata"""
-    variant_names: set[str] = set()
+    distinct_variants: dict[str, str] = {}
+    # TODO add CollectorErrataProductVariant.repos to ProductVariant
     for brew_tag in brew_tags:
         trimmed_brew_tag, sans_container_released = brew_tag_to_et_prefixes(brew_tag)
-        variants = CollectorErrataProductVersion.objects.filter(
+        variant_data = CollectorErrataProductVersion.objects.filter(
             brew_tags__overlap=[trimmed_brew_tag, sans_container_released]
-        ).values_list("variants__name", flat=True)
-        variant_names.update(variants)
-    return list(variant_names)
+        ).values("variants__name", "variants__cpe")
+        for variant in variant_data:
+            distinct_variants[variant["variants__name"]] = variant["variants__cpe"]
+
+    return distinct_variants
+
+
+def _create_inferred_variants(variants_and_cpes: dict[str, str], product: Product, product_version: ProductVersion, product_stream: ProductStream, parent: ProductNode) -> None:
+    for variant, cpe in variants_and_cpes.items():
+        variant, created = ProductVariant.objects.update_or_create(
+            name=variant,
+            defaults={
+                "cpe": cpe,
+                "products": product,
+                "productversions": product_version,
+                "productstreams": product_stream
+            }
+        )
```

Comment:
Should this be a get_or_create() here? Otherwise we'll find the existing Variant, which might be linked to many streams, and update it so that it's only linked to one stream (the one given).
```suggestion
        variant, created = ProductVariant.objects.get_or_create(
            name=variant,
            defaults={
                "cpe": cpe,
                "products": product,
                "productversions": product_version,
                "productstreams": product_stream,
            }
        )
        if not created:
            # Atomically link this existing Variant to some new Products / Versions / Streams
            # Harmless and will not add duplicates if any models given above were already linked
            variant.products.add(product)
            variant.productversions.add(product_version)
            variant.productstreams.add(product_stream)
```

Reply:
In the next PR which adds a many-to-many relationship between ProductVariant and ProductStream I've allowed the productstreams field of ProductVariant to be null so that we can avoid specifying those stream relationships up front and instead create the Product graph first, then populate the productstreams relationship of variant. Using update_or_create will allow the cpes to be updated, although I think that should never happen.